tests/unit/acquisition/function/test_active_learning.py

```diff
@@ -0,0 +1,236 @@
+        (tf.linspace([[-10.0, -10.0]], [[10.0, 10.0]], 5), tf.constant([5, 1])),
+    ],
+)
+@pytest.mark.parametrize("delta", [1, 2])
+def test_expected_feasibility_returns_correct_shape(
+    at: TensorType, acquisition_shape: TensorType, delta: int
+) -> None:
+    threshold = 1
+    alpha = 1
+    model = QuadraticMeanAndRBFKernel()
+    acq_fn = ExpectedFeasibility(threshold, alpha, delta).prepare_acquisition_function(model)
+    npt.assert_array_equal(acq_fn(at).shape, acquisition_shape)
+
+
+@pytest.mark.parametrize("delta", [1, 2])
+def test_expected_feasibility_builder_updates_without_retracing(delta: int) -> None:
+    threshold = 1
+    alpha = 1
+
+    model = QuadraticMeanAndRBFKernel()
+    builder = ExpectedFeasibility(threshold, alpha, delta)
+    acq_fn = builder.prepare_acquisition_function(model)
+    assert acq_fn._get_tracing_count() == 0  # type: ignore
+
+    query_at = tf.linspace([[-10]], [[10]], 100)
+    expected = bichon_ranjan_criterion(model, threshold, alpha, delta)(query_at)
+    npt.assert_array_almost_equal(acq_fn(query_at), expected)
+    assert acq_fn._get_tracing_count() == 1  # type: ignore
+
+    up_acq_fn = builder.update_acquisition_function(acq_fn, model)
+    assert up_acq_fn == acq_fn
+
+    npt.assert_array_almost_equal(acq_fn(query_at), expected)
+    assert acq_fn._get_tracing_count() == 1  # type: ignore
+
+
+@pytest.mark.parametrize("shape", various_shapes() - {()})
+def test_expected_feasibility_builder_raises_on_non_scalar_threshold(
+    shape: ShapeLike,
+) -> None:
+    threshold, alpha, delta = tf.ones(shape), tf.ones(shape), tf.ones(shape)
+
+    with pytest.raises(TF_DEBUGGING_ERROR_TYPES):
+        ExpectedFeasibility(threshold, 1, 1)
+
+    with pytest.raises(TF_DEBUGGING_ERROR_TYPES):
+        ExpectedFeasibility(1, alpha, 1)
+
+    with pytest.raises(TF_DEBUGGING_ERROR_TYPES):
+        ExpectedFeasibility(1, 1, delta)
+
+
+@pytest.mark.parametrize("alpha", [0.0, -1.0])
+def test_expected_feasibility_builder_raises_on_non_positive_alpha(alpha: float) -> None:
+
+    with pytest.raises(TF_DEBUGGING_ERROR_TYPES):
+        ExpectedFeasibility(1, alpha, 1)
+
+
+@pytest.mark.parametrize("delta", [-1, 0, 1.5, 3])
+def test_expected_feasibility_raises_for_invalid_delta(delta: int) -> None:
+    with pytest.raises(TF_DEBUGGING_ERROR_TYPES):
+        ExpectedFeasibility(1, 1, delta)
+
+
+@pytest.mark.parametrize("delta", [1, 2])
+@pytest.mark.parametrize("at", [tf.constant([[0.0], [1.0]]), tf.constant([[[0.0], [1.0]]])])
+def test_expected_feasibility_raises_for_invalid_batch_size(at: TensorType, delta: int) -> None:
+    ef = bichon_ranjan_criterion(QuadraticMeanAndRBFKernel(), 1, 1, delta)
+
+    with pytest.raises(TF_DEBUGGING_ERROR_TYPES):
+        ef(at)
+
+
+@pytest.mark.parametrize(
+    "threshold, at",
+    [
+        (0.0, tf.constant([[0.0]])),
+        (2.0, tf.constant([[1.0]])),
+        (-0.25, tf.constant([[-0.5]])),
+    ],
+)
+@pytest.mark.parametrize("delta", [1])
```

Comment:
the test below doesnt work for Ranjan criterion (delta=2) and I'm not sure why, I have followed Bect 2012 paper in implementing both criteria (https://arxiv.org/pdf/1009.5177.pdf;Sequential, eq 32-34), for testing here I compute the expectation from the proposition 4 from which those equations were derived, hence I think it should work for both Bichon and Ranjan. I think I have implemented all equations correctly, if you will agree with that then perhaps the derivation of eq 34 in the paper might be incorrect.

A potential source might be that Ranjan had an erratum regarding these equations (https://www.tandfonline.com/doi/abs/10.1198/TECH.2011.10192). Now, Bect paper does mention something along these lines as well , but perhaps its not that error (on pg 17 they say: "In the case δ = 2, we have found and corrected a mistake in the computations of Ranjan et al. (2008, Eq. (8) and Appendix B").